### PR TITLE
[SPARK-29327][MLLIB]Support specifying features via multiple columns

### PR DESCRIPTION
**Pull request by firestarman**
### What changes were proposed in this pull request?
This PR will mainly introduce two new APIs in both “Predictor” and “PredictionModel”, and a new parameter named “featuresCols” in “sharedParams”.
**APIs:**
`def setFeaturesCol(value: Array[String]): M = ...`
`protected def isSupportMultiColumnsForFeatures: Boolean = false`
**Parameter:**
`final val featuresCols: StringArrayParam = new StringArrayParam(this, "featuresCols",   ...)`

### Why are the changes needed?
There are always more features than one in a classification/regression task, however the current API to specify features columns in Predictor of Spark MLLib only supports one single column, which requires users to assemble the multiple features columns into a "org.apache.spark.ml.linalg.Vector" before fitting to Spark ML pipeline. So users have to spend some extra time to search and learn other transformers that can do this vectorization thing, then pick one in their applications. It looks like somewhat complicated.

With this PR, things become much more simple to run a machine learning task. Users no longer need the vectorization step for features columns, just specify these features columns by our new API, then fit the loaded the dataset directly to Spark ML Pipeline. It not only makes the ML application code more concise, but also saves a lot of time to learn other transformers for vectorization to run a classification/regression task , especially for novices at Spark ML.

### Does this PR introduce any user-facing change?
Yes, it is adding a new API `setFeaturesCol(value: Array[String]` for users to specify multiple columns for features directly, without assembling them into a vector.

### How was this patch tested?
Added unit test, and tested implementation on XGBoost-Spark.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
I don't know about MLLIB but this is clearly not a minor change, so leaving general comment to guide on contribution.

Could you please file an issue on JIRA and update the title of your PR? You're encouraged to take a look at the doc how to contribute via http://spark.apache.org/contributing.html

Thanks!

**Comment by firestarman** [comment]
> I don't know about MLLIB but this is clearly not a minor change, so leaving general comment to guide on contribution.
> 
> Could you please file an issue on JIRA and update the title of your PR? You're encouraged to take a look at the doc how to contribute via http://spark.apache.org/contributing.html
> 
> Thanks!

Thanks for reivew, filed an issue and updated the title.

**Comment by tgravescs** [comment]
ok to test

File: mllib/src/test/scala/org/apache/spark/ml/PredictorSuite.scala [hunk 1]
```diff
@@ -55,14 +55,50 @@ class PredictorSuite extends SparkFunSuite with MLlibTestSparkContext {
...
+        col("feature1").cast(StringType), col("feature2"), col("feature3")))
+    }
+
+    val df2 = df.toDF("label", "weight", "features", "feature2", "feature3")
+    // Should fail due to missing "feature1" in schema
+    intercept[IllegalArgumentException] {
+      predictor.setFeaturesCol(Array("feature1", "feature2", "feature3")).fit(df2)
+    }
+
+    // Should fail due to wrong type in schema for single column of features
```
Comment (tgravescs) [inline mllib/src/test/scala/org/apache/spark/ml/PredictorSuite.scala]: I'm don't understand what this one is testing.  what type is wrong here?  feature2 and feature3 look like ints

**Comment by tgravescs** [comment]
Overall change makes sense to me, had one minor question on the test.  It would be nice to get someone more familiar with ML lib stuff to review though.  @WeichenXu123  are you more familiar with this code?

File: mllib/src/test/scala/org/apache/spark/ml/PredictorSuite.scala (hunk 1 above)
Comment (firestarman) [reply mllib/src/test/scala/org/apache/spark/ml/PredictorSuite.scala]: This is for the case users are specifying both single column and multiple columns for features. `features` (The default value) now is used as the single column name so its type should be Vector, but actually is Int .

Comment (tgravescs) [reply mllib/src/test/scala/org/apache/spark/ml/PredictorSuite.scala]: oh, I see, so can you set the original setFeatureCol(String) to "" if you wanted to use "features" in the setFeaturesCol(Array)?  It looks like it but wanted to double check.

Can you update the comment to say something like your explanation.

Comment (firestarman) [reply mllib/src/test/scala/org/apache/spark/ml/PredictorSuite.scala]: Thanks for review.  Updated the comments
Actually that's expected. I mean only the names ("feature2", "feature3") passed into `setFeaturesCol(Array)` are wanted to use as multiple columns. But "features" is provided in "df2" schema, equal to the default value of the single column name (just like calling `setFeaturesCol("features")`). Then my current design supposes users are trying to use both single column and multiple columns, and does type check for both of them. As said above ,"features" now is used as single column, and should be "Vector" but actually "Int", so the test fails.

**Review by tgravescs — APPROVED** [review]

**Comment by tgravescs** [comment]
@zhengruifeng saw you doing some multi-column work in another PR, pinging in case you are familiar with this code as well

**Comment by tgravescs** [comment]
test this please

File: mllib/src/main/scala/org/apache/spark/ml/Predictor.scala [hunk 2]
```diff
@@ -127,6 +136,27 @@ abstract class Predictor[
...
+    set(featuresCols, value).asInstanceOf[Learner]
+  }
+
+  /**
+   * Children classes should override this to tell whether supporting multiple columns
+   * for features, meaning to fetch features columns names from parameter 'featuresCols'.
+   *
+   * @return Whether this predictor support multiple columns for features
+   */
+  protected def isSupportMultiColumnsForFeatures: Boolean = false
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/Predictor.scala]: I do not see the impl that support fit/transform in Predictor/PredictionModel, it seems that in this PR all algs in ML do not impl this new fit/transform with more features cols?

**Review by zhengruifeng — COMMENTED** [review]
I guess that only benefit is skip vectors' concatenating? 
Is it really worthwhile or is there any other famous ml libraries (sklearn/xgboost) that support this feature?

**Comment by tgravescs** [comment]
Yes the benefit is you don't have to create the vector, you can just pass in the multiple feature columns directly. An easier API to use.  It is being used with XGBoost.

**Comment by tgravescs** [comment]
@zhengruifeng  do you have specific concerns?

**Comment by srowen** [comment]
We already have a long-established component for this, VectorAssembler. One way or the other the assembly has to be done. I get that a utility setFeatures method saves a little code, but, it means changing all the implementations, while we still can't remove VectorAssembler. I'm not sure it's worth it.

**Comment by tgravescs** [comment]
right it can be done today, but this makes it  simpler to use and possibly more performant.  I think it especially helps the cases where your data is already in columnar format.  VectorAssembler has to make a pass over the data and merge multiple columns.   Many ML algorithms prefer columnar data and this allows the algorithm to determine what it wants to do with the columns. if it doesn't need to merge them, that is just extra processing.   If it already knows the format it can take advantage of that.

In this case you actually don't have to change the implementations until you want to, which is what this PR does.  Its just introducing the api, xgboost is going to use it, then other algo's can support as they want to.

**Comment by srowen** [comment]
The xgboost implementation can also just do this if desired, take n columns instead of 1 vector-valued one. Is this a real possibility or just theoretical? right now this doesn't take advantage of it anywhere in Spark ML; what would be an example?

**Comment by tgravescs** [comment]
What do you mean by can take n columns instead of 1?  The api currently allows for one parameter.

yes its being used in practice in XGBoost on GPU.   Any of the algorithms could either do the vector assembly internally (thus making it less complex for the user), or if the algorithm knows what to do with the feature columns it can skip the creation of the vector altogether.   We had some CPU code for XGboost that did this as well.  @firestarman can you provide an example.

**Comment by srowen** [comment]
You're suggesting that some implementation would prefer to be passed n columns, rather than 1 vector column of n values, because it might be more optimal in some cases. I'm asking what the use case is, either in theory or in practice. Are you saying the xgboost impl already effectively allows this?

**Comment by tgravescs** [comment]
yes we have an actual implementation that is using this for XGBoost running on GPUs.  This version takes advantage of the columnar layout.  @firestarman should have the code for the XGBoost CPU side as well that could side step creating the vector, which he should be able to point us to.

 We thought it would be a good api addition for convenience and allow the  benefits I stated above.

**Comment by srowen** [comment]
Sure, but xgboost can do that itself without adding an API that isn't (yet) otherwise used by Spark MLlib. I guess I'd wait until there's more than 1 implementation that can take advantage, to make it an API method for everyone. How does this improve xgboost, out of curiosity -- is this in the repo already?

**Comment by tgravescs** [comment]
Yes we have an external XGBoost repo you should be able to see the gpu code @firestarman  can you point to it and explain a bit?    I get what your saying but I also think sometimes its the chicken and egg problem and other external algorithms outside mllib could benefit but don't want to hassle with modifying spark. we would have to investigate if other mllib algo's would be able to skip the vector assembly or if this api would just be an easier to use api and hide the vector creation from the user.

**Comment by zhengruifeng** [comment]
> VectorAssembler has to make a pass over the data and merge multiple columns.
`VectorAssembler` only trigger a `first()` job to get the sizes of input vectors.

> Many ML algorithms prefer columnar data and this allows the algorithm to determine what it wants to do with the columns.
Do you mean column-based parallelism used in distributed tree building? Such function is not exposed to end users, and what you need to do is only to set params like `(..., updater=distcol)`.
If some alg will benefit from column-based parallelism, I guess it is better to split the features internally. No alg in MLLibs is designed to fit/transform with column-based datasets for now, so I do not prefer to add this feature.

> It is being used with XGBoost.
I cannot find any related docs in [XGBoost Parameters](https://xgboost.readthedocs.io/en/latest/parameter.html#xgboost-parameters). Could you please provide a link for this?

**Comment by firestarman** [comment]
> Yes we have an external XGBoost repo you should be able to see the gpu code @firestarman can you point to it and explain a bit? 

Here are the links of gpu code
1) Column names => indices 
https://github.com/rapidsai/xgboost/blob/rapids-spark/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala#L484
2) indices => handles of column data
https://github.com/rapidsai/xgboost/blob/rapids-spark/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala#L917

You can see features column names are used directly to transform to the handles of columns data,  then build the columnar based DMatrix for XGBoost.

**Comment by firestarman** [comment]
> `VectorAssembler` only trigger a `first()` job to get the sizes of input vectors.
> And I think we can futhuer optimize it so that this `first()` job is not triggered when all columns are double values.
> 
> 
> Do you mean column-based parallelism used in distributed tree building? Such function is not exposed to end users, and what you need to do is only to set params like `(..., updater=distcol)`.
> If some alg will benefit from column-based parallelism, I guess it is better to split the features internally.

Thanks for review.
Actually this is not only for alg that benefit from column-based parallelism, but also for spark ML application users, with this PR and its implemention in MLLibs, they don't need `VectorAssembler` any more for ML features column things, but simply and directly set feature column names to `Predictor` instead, just like what you do with `label` and `prediction`.
What's more, the original column names are required for MLLibs to parse the info of the column-based datasets. But currently there is no API to save the original column names in a `Predictor` .


> No alg in MLLibs is designed to fit/transform with column-based datasets for now, so I do not prefer to add this feature. I cannot find any related docs in [XGBoost Parameters](https://xgboost.readthedocs.io/en/latest/parameter.html#xgboost-parameters). Could you please provide a link for this?

XGBoost should support column-based datasets now.  you can find clues from code or PRs, like 
https://github.com/dmlc/xgboost/pull/4816
https://github.com/dmlc/xgboost/pull/4850
https://github.com/dmlc/xgboost/pull/4902

One more thing, would it be more reasonable to define the APIs firstly in Spakr ML interface for requirements from users, then have them implemented in Spark MLLibs and other ML implementations ?

**Comment by tgravescs** [comment]
@zhengruifeng

**Comment by tgravescs** [comment]
Is the lack of response because people are ok with the explanation?   
There are also lots of algorithms outside the spark core ml lib code that use these classes and so we were thinking this would be convenience method that doesn't add much overhead to the spark core code.  It seems like pure overhead to create the VectorAssembler if the algorithm doesn't need it.

**Comment by srowen** [comment]
So, xgboost's impl can already do this. It doesn't need to be an API in Spark. If Spark implementations can benefit from it, then it makes sense to add this API. But if there aren't any other impls, I don't think it adds anything, and adds a second way to do something you can do already. I'd rather not until it buys more. 

Am I missing something... why not just expose this in xgboost right now?

**Comment by zhengruifeng** [comment]
If current algs in MLLIB will not benefit from this design (execpt skip `VectorAssembler`), I do not prefer it. SKLearn is also widely used, it do not support this feature. 

> There are also lots of algorithms outside the spark core ml lib code that use these classes and so we were thinking this would be convenience method that doesn't add much overhead to the spark core code.

Yes, but I think current design in MLLIB do not pose an obstacle to impl third-party libs.

**Comment by tgravescs** [comment]
Right but if its not in Spark then multiple algos/libs can't use it and will likely re-invent the same thing. People are not going to see it in xgboost and then decide to add it to Spark.  This is why I encouraged Liangcai to propose it.  To me its useful to XGBoost so the question is then could it be useful to other algorithms in the future. To me I thought that answer was yes for the following reasons:

- Simpler, more user friendly API
  - more obvious to user how to send in multiple features
  - user doesn't have to create VectorAssembler
  - algorithm could behind the scene create the Vector if required
  - algorithm could skip if its not required
  - don't lose the information about original column names
  - give the algorithm more flexibility
- more performant if can skip the VectorAssembler
- If GPU algorithm could be a lot more performant if data is already columnar

Then on top of that, its really not much code so I didn't think it introduced much maintenance on the Spark side.  We could mark it with Experimental as well and remove it if really not used by anyone else.

I'm by no means an expert in all ML algorithms, so if that doesn't make sense and you don't think matches way everyone is doing it then we can just leave it in XGboost then.

**Comment by tgravescs** [comment]
@firestarman  please close this
